src/clr-angular/data/tree-view/tree-focus-manager.service.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable } from '@angular/core';
+import { Observable, Subject } from 'rxjs';
+import { TreeNodeModel } from './models/tree-node.model';
+
+@Injectable()
+export class TreeFocusManagerService<T> {
+  public rootNodeModels: TreeNodeModel<T>[];
+
+  private focusedNodeId: string;
+
+  private _focusRequest: Subject<string> = new Subject();
+
+  private _focusChange: Subject<string> = new Subject();
+
+  get focusRequest(): Observable<string> {
+    return this._focusRequest.asObservable();
+  }
+
+  get focusChange(): Observable<string> {
+    return this._focusChange.asObservable();
+  }
+
+  private findParentChildren(model: TreeNodeModel<T>): TreeNodeModel<T>[] {
+    if (model.parent) {
+      return model.parent.children;
+    } else {
+      return this.rootNodeModels;
+    }
+  }
+
+  private findLastVisibleInNode(model: TreeNodeModel<T>): TreeNodeModel<T> {
+    if (!model) {
+      return;
+    }
+    if (model.expanded && model.children.length > 0) {
+      const children = model.children;
+      const lastChild = children[children.length - 1];
+      return this.findLastVisibleInNode(lastChild);
+    } else {
+      return model;
+    }
+  }
+
+  private findNextInParentChildren(model: TreeNodeModel<T>): TreeNodeModel<T> {
```

Comment:
Similar comment on naming here.. something like `findNextFocusableNode` is probably clearer. The logic looks good. 👍